### PR TITLE
Part1

### PR DESCRIPTION
**Pull request by just2high**

### CODE REVIEW

File: Character.cpp [hunk 1]
```diff
@@ -27,6 +28,9 @@ void Character::attack( Character& other )
     isDefending = false;
     std::cout << getName() << " has attacked " << other.getName() << std::endl;
     //subtract attackDamage from other->hitPoints
+
+    other.hitPoints -= attackDamage;
+
```
Comment (matkatmusic) [inline Character.cpp]: - [x] remove this line.
- [x] Look at what `takeDamage()` does.

File: Character.cpp [hunk 2]
```diff
@@ -87,31 +91,51 @@ int Character::takeDamage(int damage)
...
+            attackDamage = ceil(attackDamage * 1.1);
+
+        if( initialArmorLevel != nullptr; armor < *initialArmorLevel)
+            armor = ceil(*initialArmorLevel * 1.1);
+        else
+            armor = ceil(armor * 1.1);
+
+        *initialHitPoints = hitPoints;
+        *initialArmorLevel = armor;
+        *initialAttackDamage = attackDamage;
```
Comment (matkatmusic) [inline Character.cpp]: if all 6 of these member variables hold the same type, see if you can write a member function that will D.R.Y. all of this. 
- [x] find a D.R.Y refactor solution.

File: DefensiveItem.cpp [hunk 3]
```diff
@@ -1,12 +1,14 @@
 #include "DefensiveItem.h"
-#include "Dwarf.h"
-#include "Paladin.h"
-#include "DragonSlayer.h"
-#include "Dragon.h"
-#include <assert>
+// #include "Dwarf.h"
+// #include "Paladin.h"
+// #include "DragonSlayer.h"
+// #include "Dragon.h"
```
Comment (matkatmusic) [inline DefensiveItem.cpp]: - [x] remove dead code/commented code.   keep it clean!

File: DragonSlayer.cpp [hunk 4]
```diff
@@ -1,27 +1,48 @@
...
     if( auto* dragon = dynamic_cast<Dragon*>(&other) )
     {
-        assert(false);
+        //assert(false);
         //DragonSlayers get a 10x boost when attacking dragons, from their attack item.
         //so they should USE their attack item before attacking the dragon... 
         //
+
+        useAttackItem(this, &dragonSword);
+
```
Comment (matkatmusic) [inline DragonSlayer.cpp]: call the member function on your attack item.
remember, items are 1-time-use only, so you need to reset it after use.  
look in the Character class for how the other item types get used/reset.
- [x] revise.

File: DragonSlayer.cpp [hunk 5]
```diff
@@ -1,27 +1,48 @@
...
+        useAttackItem(this, &dragonSword);
+        
         while( dragon->getHP() > 0 )
         {
           dragon->takeDamage(attackDamage);
         }
+        //reset attack damage after killing dragon
+        attackDamage -= dragonSword.getBoost() * 10;
+
+        std::cout << getName() << "'s frenzy is over.  Damage returns to: " << attackDamage << std::endl;
```
Comment (matkatmusic) [inline DragonSlayer.cpp]: Good point!  item effects should expire.

File: DragonSlayer.h [hunk 6]
```diff
@@ -1,8 +1,19 @@
...
 {
+    DragonSlayer(std::string name_, int hp, int armor);
+    ~DragonSlayer () override;
+
+    const std::string& getName() override;
+    std::string getStats() override;
+
     void attack(Character& other) override;
+
+    AttackItem dragonSword;
```
Comment (matkatmusic) [inline DragonSlayer.h]: This needs to be resettable, like the other items. 
- [x] store in a `unique_ptr<>`

File: Dwarf.cpp [hunk 7]
```diff
@@ -1,7 +1,27 @@
...
+std::string Dwarf::getStats()
+{
+    return getCharacterStats(this);
+}
+
+void Dwarf::attack( Character& other )
+{
+    std::cout << getName() << " is attacking " << other.getName() << "!\n";
+    Character::attack( other );
+}
```
Comment (matkatmusic) [inline Dwarf.cpp]: - [] remove this overridden function.  base class implementation does the job.

File: Dwarf.h [hunk 8]
```diff
@@ -1,7 +1,16 @@
...
 
 struct Dwarf : Character
 {
+    Dwarf( std::string name_, int hp, int armor );
+    ~Dwarf() override;
+
+    const std::string& getName() override;
+    std::string getStats() override;
+    
+    void attack( Character& other ) override;
```
Comment (matkatmusic) [inline Dwarf.h]: - [ ] remove this declaration

File: Paladin.cpp [hunk 9]
```diff
@@ -1,7 +1,27 @@
...
+std::string Paladin::getStats()
+{
+    return getCharacterStats(this);
+}
+
+void Paladin::attack( Character& other )
+{
+    std::cout << getName() << " is attacking " << other.getName() << "!\n";
+    Character::attack( other );
+}
```
Comment (matkatmusic) [inline Paladin.cpp]: - [ ] remove this function.  base class implementation already does this.

File: Paladin.h [hunk 10]
```diff
@@ -1,7 +1,16 @@
...
 
 struct Paladin : Character
 {
+    Paladin(std::string name_, int hp, int armor);
+    ~Paladin() override;
+
+    const std::string& getName() override;
+    std::string getStats() override;
+    
+    void attack( Character& other ) override;
```
Comment (matkatmusic) [inline Paladin.h]: - [ ] remove this declaration

File: Utility.cpp [hunk 11]
```diff
@@ -43,66 +42,67 @@ std::string getCharacterStats(Character* ch)
...
-    else if( auto* ch = dynamic_cast<DragonSlayer*>(character))
+    else if( auto* dragonSlayer = dynamic_cast<DragonSlayer*>(character))
     {
-        assert(false);
+        //assert(false);
         //DragonSlayers get a 10x boost when attacking dragons, from their attack item.
         //so their attack item should boost their attack damage by a factor of 10
         //this means you need to GET the attack damage, multiply it by the item's boost, and BOOST the attackDamage with that multiplied value.  
         //check Character.h for available member functions you can use.
+        dragonSlayer->boostAttackDamage(item->getBoost() * 10);
```
Comment (matkatmusic) [inline Utility.cpp]: the item's boost is already 10.  you are not getting the attack damage.
- [x] get attack damage, boost attack damage with item's boost value.

**Review by matkatmusic — CHANGES_REQUESTED** [review]
Please click "resolve conversation" for each item after you implement the change.

File: Character.cpp [hunk 12]
```diff
@@ -87,31 +91,48 @@ int Character::takeDamage(int damage)
...
+
+void Character::levelUp( int& currentValue, int initialValue )
+{
+    if( currentValue < initialValue )
+        currentValue = ceil(initialValue * 1.1);
+    else
+        currentValue = ceil(currentValue * 1.1);
+
+    initialValue = currentValue;
+}
```
Comment (matkatmusic) [inline Character.cpp]: Good, but you need to pass `initialValue` by reference, or return currentValue and assign it to whatever stat you're updating.  Right now, you are not updating whatever you're passing to `initialValue` which means line 103 is not doing anything.
- [ ] fix param type.

File: DragonSlayer.cpp [hunk 13]
```diff
@@ -1,27 +1,49 @@
...
         //
+        dragonSword->use(this);
+        dragonSword.reset();
+
         while( dragon->getHP() > 0 )
         {
           dragon->takeDamage(attackDamage);
         }
+        //reset attack damage after killing dragon
+        attackDamage = this->getInitialAttackDamage();
```
Comment (matkatmusic) [inline DragonSlayer.cpp]: `initialAttackDamage` is actually protected.  you don't need a member function to access it in derived classes.
- [ ] remove getter usage.  assign directly.

File: Character.h [hunk 14]
```diff
@@ -30,6 +30,7 @@ struct Character
     int getHP() const { return hitPoints; }
     int getArmorLevel() const { return armor; }
     int getAttackDamage() const { return attackDamage; }
+    int getInitialAttackDamage() const { return *initialAttackDamage; }
```
Comment (matkatmusic) [inline Character.h:33]: - [ ] remove this.  `initialAttackDamage` is protected.  derived classes don't need a member function to access it.

File: DragonSlayer.h [hunk 15]
```diff
@@ -1,8 +1,20 @@
...
 {
+    DragonSlayer(std::string name_, int hp, int armor);
+    ~DragonSlayer () override;
+
+    const std::string& getName() override;
+    std::string getStats() override;
+
     void attack(Character& other) override;
+
+    std::unique_ptr<AttackItem> dragonSword { new AttackItem() };
```
Comment (matkatmusic) [inline DragonSlayer.h]: This should be private.

**Review by matkatmusic — CHANGES_REQUESTED** [review]
Close.   
via Slack: can you explain why you have those empty destructors in your derived classes

File: DragonSlayer.cpp (hunk 13 above)
Comment (just2high) [reply DragonSlayer.cpp]: whenever I try to access `initialAttackDamage ` directly I get a "Can't access private member variable" error

File: Character.h (hunk 14 above)
Comment (just2high) [reply Character.h:33]:
```
protected:
    std::vector<std::unique_ptr<Item>> defensiveItems;
    std::vector<std::unique_ptr<Item>> helpfulItems;
    int hitPoints, armor;
    int attackDamage;
    bool isDefending = false;
private:
    std::unique_ptr<int> initialHitPoints, initialArmorLevel, initialAttackDamage;
    
    void attackInternal(Character& other);
};
```
`  It's private, not protected, what I find via google is private members of base class are inaccessible to derived class, is that not true?`

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**